src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/EETypeNode.cs

```diff
@@ -457,5 +501,128 @@ private void OutputNullableTypeParameter(NodeFactory factory, ref ObjectDataBuil
+        /// at least one optional field was set.
+        /// </summary>
+        private void ComputeOptionalEETypeFields(NodeFactory factory)
+        {
+            // Todo: DispatchMap table index when we support interface dispatch maps
+            ComputeRareFlags();
+            ComputeNullableValueOffset();
+            ComputeICastableVirtualMethodSlots(factory);
+            ComputeValueTypeFieldPadding();
+        }
+
+        void ComputeRareFlags()
+        {
+            uint flags = 0;
+
+            if (_type.IsNullable)
+            {
+                flags |= (uint)EETypeRareFlags.IsNullableFlag;
+            }
+
+            if (_type.HasStaticConstructor)
+            {
+                flags |= (uint)EETypeRareFlags.HasCctorFlag;
+            }
+            
+            if (TypeRequiresAlign8(_type))
+            {
+                flags |= (uint)EETypeRareFlags.RequiresAlign8Flag;
+            }
+
+            if (_type is DefType && ((DefType)_type).IsHFA())
+            {
+                flags |= (uint)EETypeRareFlags.IsHFAFlag;
+            }
+
+            if (flags != 0)
+            {
+                _optionalFieldsBuilder.SetFieldValue(EETypeOptionalFieldsElement.RareFlags, flags);
+            }
+        }
+
+        /// <summary>
+        /// To support boxing / unboxing, the offset of the value field of a Nullable type is recorded on the EEType.
+        /// This is variable according to the alignment requirements of the Nullable&lt;T&gt; type parameter.
+        /// </summary>
+        void ComputeNullableValueOffset()
+        {
+            if (!_type.IsNullable)
+                return;
+
+            var field = _type.GetField("value");
+
+            // Ensure the definition of Nullable<T> didn't change on us
+            Debug.Assert(field != null);
+
+            // In the definition of Nullable<T>, the first field should be the boolean representing "hasValue"
+            Debug.Assert(field.Offset > 0);
+
+            // The contract with the runtime states the Nullable value offset is stored with the boolean "hasValue" size subtracted
+            // to get a small encoding size win.
+            _optionalFieldsBuilder.SetFieldValue(EETypeOptionalFieldsElement.NullableValueOffset, (uint)field.Offset - 1);
+        }
+
+        /// <summary>
+        /// ICastable is a special interface whose two methods are not invoked using regular interface dispatch.
+        /// Instead, their VTable slots are recorded on the EEType of an object implementing ICastable and are
+        /// called directly.
+        /// </summary>
+        void ComputeICastableVirtualMethodSlots(NodeFactory factory)
+        {
+            // TODO: This method is untested (we don't support interfaces yet)
+            if (_type.IsInterface)
+                return;
+
+            foreach (DefType itf in _type.RuntimeInterfaces)
+            {
+                if (itf.Name == "System.Runtime.CompilerServices.ICastable")
```

Comment:
The `ICastable` type should be fetched from SystemModule and cached on the NodeFactory (next to HelperEntrypoint cache).